### PR TITLE
Relax db

### PR DESCRIPTION
**Pull request by josenavas**
Note: this PR is made against the new `relax-md-req`.

I'm only changing the SQL layout and the populate_test_db.sql, so people can check if there is something terribly wrong.

I'm not changing any code, so tests are expected to fail.
In the next PR, (which will be build on top of this one) I'll start fixing the metadata_template objects, so some tests start passing with the new layout :-)

### CODE REVIEW

**Comment by josenavas** [comment]
@biocore/qiita-dev This is ready for review. Note that this is going to it's own branch. Tests are expected to fail. The only thing that I'm changing in this PR is the database layout and the populate test database file, so it can be successfully added after the new changes have been made.

**Comment by josenavas** [comment]
@biocore/qiita-dev The `relax-md-req` in upstream is now based on @squirrelo's changes, so this is modifying on top of his changes. Can be this be reviewed? Note that this is only changing the database, and the idea is get some feedback in terms of "don't do this is crazy". The earlier that I get that feedback the less burden I'll have lter.

File: qiita_db/support_files/patches/python_patches/23.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+from qiita_db.metadata_template import SampleTemplate, PrepTemplate
+
+conn_handler = SQLConnectionHandler()
+
+# Relaxing constraints on the Sample Template
+# Retrieve all study ids that have a sample template
+sql = """SELECT DISTINCT study_id from qiita.required_sample_info"""
+study_ids = set(s[0] for s in conn_handler.execute_fetchall(sql))
+
+queue_name = "PATCH_21"
```
Comment (antgonza) [inline qiita_db/support_files/patches/python_patches/23.py]: queue name should be 23, right? Not blocking.

Comment (josenavas) [inline qiita_db/support_files/patches/python_patches/23.py]: Good catch. Changed to avoid future confusions.

**Comment by antgonza** [comment]
Looks fine but it seems like all these changes can be done directly in a sql and not in python ...

**Comment by josenavas** [comment]
I can investigate on how to do everything on SQL. I'll dedicate a short period of time, otherwise I'm not going to meet the deadline. If you have a suggestion it will likely help me a lot to reduce development burden.

**Comment by antgonza** [comment]
I do not think is blocking, just curious why this choice?

**Comment by josenavas** [comment]
Shorter development time, I'm still getting used to more complex SQL queries, although I'm pretty exited on taking advantage of SQL.

**Comment by antgonza** [comment]
OK :+1: as we do not have any rules about this what should go in a sql or py patch.

**Comment by josenavas** [comment]
Great thanks

File: qiita_db/support_files/patches/python_patches/23.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            DROP COLUMN host_subject_id,
+            DROP COLUMN description,
+            DROP COLUMN latitude,
+            DROP COLUMN longitude"""
+conn_handler.add_to_queue(queue_name, sql)
+
+# Since that table no longer stores required metadata,
+# we are going to rename it
+sql = """ALTER TABLE qiita.required_sample_info
+            RENAME TO study_sample"""
```
Comment (wasade) [inline qiita_db/support_files/patches/python_patches/23.py]: Is the schema implicit?

File: qiita_db/support_files/patches/python_patches/23.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+                WHERE prep_template_id=%s"""
+
+sql_cols = """INSERT INTO qiita.prep_columns
+                (prep_template_id, column_name, column_type)
+              VALUES (%s, 'center_name', 'varchar'),
+                     (%s, 'center_project_name', 'varchar'),
+                     (%s, 'emp_status', 'varchar')"""
+
+sql_alter = """
+ALTER TABLE qiita.{0} ADD COLUMN center_name varchar;
```
Comment (squirrelo) [inline qiita_db/support_files/patches/python_patches/23.py]: Can also do add column in one go with commas.

File: qiita_db/support_files/patches/python_patches/23.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+                     (%s, 'sample_type', 'varchar'),
+                     (%s, 'collection_timestamp', 'timestamp'),
+                     (%s, 'host_subject_id', 'varchar'),
+                     (%s, 'description', 'varchar'),
+                     (%s, 'latitude', 'float8'),
+                     (%s, 'longitude', 'float8'),
+                     (%s, 'required_sample_info_status', 'varchar')"""
+
+sql_alter = """
+ALTER TABLE qiita.{0} ADD COLUMN physical_specimen_location varchar;
```
Comment (squirrelo) [inline qiita_db/support_files/patches/python_patches/23.py]: again, commas work for add column.

File: qiita_db/support_files/patches/python_patches/23.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+                    host_subject_id=%s,
+                    description=%s,
+                    latitude=%s,
+                    longitude=%s,
+                    required_sample_info_status=%s
+                WHERE sample_id=%s"""
+
+for s_id in study_ids:
+    # Get the data from the requried_sample_info
+    data = conn_handler.execute_fetchall(sql_select, (s_id,))
```
Comment (wasade) [inline qiita_db/support_files/patches/python_patches/23.py]: This `execute_fetchall` and the others are outside of the queues, so the data the return are not assured to reflect the state of the database when the transaction block is executed.

File: qiita_db/support_files/patches/python_patches/23.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+sql = """ALTER TABLE qiita.required_sample_info
+            RENAME TO study_sample"""
+conn_handler.add_to_queue(queue_name, sql)
+
+conn_handler.add_to_queue(
+    queue_name, """DROP TABLE qiita.required_sample_info_status""")
+
+# Relaxing constrains on the prep template
+# Retrieve all prep template ids
+sql = """SELECT prep_template_id from qiita.prep_template"""
```
Comment (squirrelo) [inline qiita_db/support_files/patches/python_patches/23.py]: if you do `SELECT UNIQUE` you don't need to make it a set in the line below.

File: qiita_db/support_files/patches/python_patches/23.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+    # Get the dynamic table name
+    table_name = "prep_%d" % p_id
+
+    # Add the columns to the dynamic table
+    conn_handler.add_to_queue(queue_name, sql_alter.format(table_name))
+
+    # Add the values for each sample
+    for sample_data in data:
+        sample_data = dict(sample_data)
```
Comment (squirrelo) [inline qiita_db/support_files/patches/python_patches/23.py]: don't need explicit dict change if not returning as a dict type.

File: qiita_db/support_files/patches/python_patches/23.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+    # Get the dynamic table name
+    table_name = "sample_%d" % s_id
+
+    # Add the columns to the dynamic table
+    conn_handler.add_to_queue(queue_name, sql_alter.format(table_name))
+
+    # Add the values for each sample
+    for sample_data in data:
+        sample_data = dict(sample_data)
```
Comment (squirrelo) [inline qiita_db/support_files/patches/python_patches/23.py]: don't need explicit dict change if not returning as a dict type.

File: qiita_db/support_files/patches/python_patches/23.py [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    queue_name,
+    """ALTER TABLE qiita.common_prep_info
+        DROP COLUMN center_name,
+        DROP COLUMN center_project_name,
+        DROP COLUMN emp_status_id""")
+
+# Since that table no longer stores common prep template metadata,
+# we are going to rename it
+sql = """ALTER TABLE qiita.common_prep_info
+            RENAME TO prep_template_sample"""
```
Comment (squirrelo) [inline qiita_db/support_files/patches/python_patches/23.py]: as @wasade said above, this probably needs schema in front of the name.

**Comment by squirrelo** [comment]
Few comments on the python patch. All in all, the logic on this is extremely hard to follow and I think it would be greatly improved if moved over to straight SQL. This would also take care of @wasade's concern about the selects being outside the queue.

**Comment by wasade** [comment]
@squirrelo, agree. I think it would be dangerous to have this patch run on a database that is not absolutely quiet (as no other connections except the one executing this patch).

**Comment by josenavas** [comment]
Thanks guys, I will try to do everything in SQL, those concerns are legitimate.

File: qiita_db/support_files/patches/python_patches/23.py (hunk 2 above)
Comment (josenavas) [inline qiita_db/support_files/patches/python_patches/23.py]: I will move this to SQL but just FYI, since you are changing the name of the table you don't need the schema. Actually, if you put the schema does not work (I've tested it on my system cause I also thought it was needed)

File: qiita_db/support_files/patches/python_patches/23.py (hunk 5 above)
Comment (josenavas) [inline qiita_db/support_files/patches/python_patches/23.py]: FYI, it was done like that because the queues does not support that.

Comment (wasade) [inline qiita_db/support_files/patches/python_patches/23.py]: That would mean queues aren't appropriate here, right? 

The chance of something going wrong is low, but the risk is high, unless there's something I'm misunderstanding?

**Comment by josenavas** [comment]
@wasade @antgonza @squirrelo I moved everything to the SQL. Ready for review.

File: qiita_db/support_files/patches/23.sql [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+                physical_specimen_remaining=sv.has_physical_specimen,
+                dna_extracted=sv.has_extracted_data,
+                sample_type=sv.sample_type,
+                collection_timestamp=sv.collection_timestamp,
+                host_subject_id=sv.host_subject_id,
+                description=sv.description,
+                latitude=sv.latitude,
+                longitude=sv.longitude,
+                required_sample_info_status=sv.status
+            FROM sample_values sv
```
Comment (squirrelo) [inline qiita_db/support_files/patches/23.sql]: Why rename it her eto sv when you can use `WITH sv AS` on line 52.

File: qiita_db/support_files/patches/23.sql [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+    EXECUTE '
+        WITH sample_values AS (SELECT *
+                               FROM qiita.common_prep_info
+                                    JOIN qiita.emp_status USING (emp_status_id)
+                               WHERE prep_template_id = ' || prep_id || ')
+        UPDATE ' || dyn_table || '
+            SET center_name=sv.center_name,
+                center_project_name=sv.center_project_name,
+                emp_status=sv.emp_status
+            FROM sample_values sv
```
Comment (squirrelo) [inline qiita_db/support_files/patches/23.sql]: same as above

**Comment by squirrelo** [comment]
Python patch is broken:
```
  File "/Users/jshorens/Repositories/qiita/qiita_db/support_files/patches/python_patches/23.py", line 5, in <module>
    from qiita_db.metadata_template import SampleTemplate, PrepTemplate
  File "/Users/jshorens/Repositories/qiita/qiita_db/metadata_template/__init__.py", line 9, in <module>
    from .sample_template import SampleTemplate
  File "/Users/jshorens/Repositories/qiita/qiita_db/metadata_template/sample_template.py", line 23, in <module>
    from .prep_template import PrepTemplate
  File "/Users/jshorens/Repositories/qiita/qiita_db/metadata_template/prep_template.py", line 56, in <module>
    class PrepTemplate(MetadataTemplate):
  File "/Users/jshorens/Repositories/qiita/qiita_db/metadata_template/prep_template.py", line 70, in PrepTemplate
    id_cols_handlers = {'emp_status_id': get_emp_status()}
  File "/Users/jshorens/Repositories/qiita/qiita_db/util.py", line 268, in get_emp_status
    return dict(con.execute_fetchall(sql))
  File "/Users/jshorens/Repositories/qiita/qiita_db/sql_connection.py", line 454, in execute_fetchall
    with self._sql_executor(sql, sql_args) as pgcursor:
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/Users/jshorens/Repositories/qiita/qiita_db/sql_connection.py", line 303, in _sql_executor
    (sql, str(sql_args), e)))
qiita_db.exceptions.QiitaDBExecutionError: 
Error running SQL query: select emp_status, emp_status_id from qiita.emp_status
ARGS: None
Error: relation "qiita.emp_status" does not exist
LINE 1: select emp_status, emp_status_id from qiita.emp_status
```

**Comment by josenavas** [comment]
Yes, python patch is broken because it relies on code that it needs to be added later. Just as a remainder, this is broken up a 44 file PR that we agreed was too hard to review. The system will not be fully functional until all PR go in the branch...

**Comment by squirrelo** [comment]
How are you testing the patch then? This needs to be complete for the patch to work at least, if it's going to be the actual patch PR.

**Comment by josenavas** [comment]
I've tested only the SQL patch, as the python patch is independent of the data in the database.

I've commented out the python patch for the time being, so I can add it back once the functionality that I need is in place. @squirrelo if you don't agree with that, I'll happy to do another 44 files PR and I hope you'll have fun reviewing it.

**Comment by squirrelo** [comment]
Since this is on a separate branch, can you please pull the python patch out of this PR and add it back in when it works, so we don't forget to test it later on.

**Comment by josenavas** [comment]
Done

**Comment by squirrelo** [comment]
:+1:
